// media-sorcerer/karma.conf.js
/* eslint-env node */
const sourcemaps = require('rollup-plugin-sourcemaps');
const babel = require('rollup-plugin-babel');
const nodeResolve = require('rollup-plugin-node-resolve');

module.exports = function(config) {
    config.set({
        frameworks: ['mocha', 'chai-as-promised', 'sinon-chai'],
        files: [
            { pattern: 'test/**/*-test.js', watched: true, included: true },
            { pattern: 'src/**/*.js', watched: true, included: false }
        ],
        preprocessors: {
            'test/**/*-test.js': ['rollup']
        },
        reporters: ['progress', 'coverage'],
        browsers: ['Electron'],
        rollupPreprocessor: {
            plugins: [
                sourcemaps(),
                babel({
                    exclude: [
                        'node_modules/**'
                    ],
                    plugins: [
                        ['istanbul', {
                            exclude: [
                                'test/**/*.js'
                            ]
                        }]
                    ]
                }),
                nodeResolve({
                    jsnext: true
                })
            ]
        },
        coverageReporter: {
            dir: 'coverage',
            reporters: [
                { type: 'html' },
                { type: 'text-summary' }
            ]
        },
        electronOpts: {
            show: false
        }
    });
};
